### DISCUSSION TITLE
Does terramate work/support git's worktree

### DISCUSSION THREAD

**Discussion by az-z** [terramate-io/terramate#808] [Q&A] [answered]
Does terramate work/support git's worktree ?

Describe the solution you'd like
worktree is used to checkout a branch, completed all the work, and deploy from this branch.

Describe alternatives you've considered
checking out a branch into the same directory is error-prone since I have a few branches at the same time.

**Comment by i4ki** [comment]
As explained in another question, Terramate does not support `worktree` but mostly because we never used it. The change to make basic features work was very small, and it was implemented here: https://github.com/mineiros-io/terramate/pull/813
We are still not sure it supports all Terramate features, so feel free to try it out and please report if something is not working well.

Thanks for bringing this to our attention!

**Comment by mariux** [comment, marked as answer]
Terramate might now just work with Git Worktrees, but we do not cover it with tests as of today. 

Please try it out and report any issues you see while using it. Use it with caution though ;)

**Reply by az-z** [reply]
@i4ki and @i4ki ,
here is what I get while attempting to execute the "plan" command in TM. 
//executing terraform plan/apply within each stack directory works as expected

```
[az@dell5000 SMSMS-52]$ terramate run plan --dry-run
2023-03-26T13:49:45-04:00 FTL error="fetching remote commit of origin/main: Git.FetchRemoteRev: git ls-remote \"origin\" \"main\" can't parse: " action=remoteDefaultCommit()
[az@dell5000 SMSMS-52]$ git --version
git version 2.39.2
[az@dell5000 SMSMS-52]$ terramate --version
0.2.16
[az@dell5000 SMSMS-52]$ terraform --version
Terraform v1.3.9
on linux_amd64

Your version of Terraform is out of date! The latest version
is 1.4.2. You can update by downloading from https://www.terraform.io/downloads.html
[az@dell5000 SMSMS-52]$ git worktree list
/home/az/Work/SMS/Infra/SMS-IAC/master    e7ffa0c [master]
/home/az/Work/SMS/Infra/SMS-IAC/SMSMS-14  c3afa85 [SMSMS-14]
/home/az/Work/SMS/Infra/SMS-IAC/SMSMS-52  aa73d1a [feature/SMSMS-52-convert-iac-to-terramate]
[az@dell5000 SMSMS-52]$ pwd
/home/az/Work/SMS/Infra/SMS-IAC/SMSMS-52
[az@dell5000 SMSMS-52]$ git status
On branch feature/SMSMS-52-convert-iac-to-terramate
Your branch is ahead of 'origin/feature/SMSMS-52-convert-iac-to-terramate' by 1 commit.
  (use "git push" to publish your local commits)

nothing to commit, working tree clean
```

**Reply by i4ki** [reply]
Hey @az-z 

It looks like your default branch is `master` but Terramate uses `main` by default.
This can be configured by the `terramate.config.git.default_branch` in the root configuration. 
Example:

```
terramate {
  config {
    git {
      default_branch = "master"
    }
  }
}
```

_Note: you're using `terramate run plan --dry-run` but the correct one is `terramate run -- terraform plan --dry-run` (this is not related to the error message though)_

**Comment by az-z** [comment]
Marius,
Perfect timing. I'm about to give the terra another try.
Please allow me a few days for the feedback.

**Reply by az-z** [reply]
@i4ki  and @mariux ,
 I just finished migration to TM for one of the projects. I did not encounter issues related to the support of the worktree.
Thank you _very_ much for releasing this feature.

**Reply by i4ki** [reply]
That's great!

Thanks for letting us know!
